src/std/io/bio/inline.ss

```diff
@@ -3,49 +3,61 @@
 ;;; Buffered IO inline operations
 (import :std/sugar
         :std/interface
+        ../interface
         ./types
         ./input
         ./output)
 (export #t)
 
+(defrule (is-input-buffer? reader)
+  (input-buffer? (&interface-instance-object reader)))
+(defrule (is-output-buffer? writer)
+  (output-buffer? (&interface-instance-object writer)))
+
 (defrule (&BufferedReader-read-u8-inline reader)
-  (let ()
-    (declare (not interrupts-enabled))
-    (let (bio (&interface-instance-object reader))
-      (let ((rlo (&input-buffer-rlo bio))
-            (rhi (&input-buffer-rhi bio))
-            (buf (&input-buffer-buf bio)))
-        (if (fx< rlo rhi)
-          (let ((u8 (u8vector-ref buf rlo))
-                (rlo+1 (fx+ rlo 1)))
-            (bio-input-advance! bio rlo+1 rhi)
-            u8)
-          ;; empty buffer, fall back to the method
-          (bio-read-u8 bio))))))
+  (if (is-input-buffer? reader)
+    (let ()
+      (declare (not interrupts-enabled))
+      (let (bio (&interface-instance-object reader))
+        (let ((rlo (&input-buffer-rlo bio))
+              (rhi (&input-buffer-rhi bio))
+              (buf (&input-buffer-buf bio)))
+          (if (fx< rlo rhi)
+            (let ((u8 (u8vector-ref buf rlo))
+                  (rlo+1 (fx+ rlo 1)))
+              (bio-input-advance! bio rlo+1 rhi)
+              u8)
+            ;; empty buffer, fall back to the method
+            (bio-read-u8 bio)))))
+    (&BufferedReader-read-u8 reader)))
 
 (defrule (&BufferedReader-peek-u8-inline reader)
-  (let ()
-    (declare (not interrupts-enabled))
-    (let (bio (&interface-instance-object reader))
-      (let ((rlo (&input-buffer-rlo bio))
-            (rhi (&input-buffer-rhi bio))
-            (buf (&input-buffer-buf bio)))
-        (if (fx< rlo rhi)
-          (u8vector-ref buf rlo)
-          ;; empty buffer, fall back to the method
-          (bio-peek-u8 bio))))))
+  (if (is-input-buffer? reader)
+    (let ()
+      (declare (not interrupts-enabled))
```

Comment:
Is (not interrupts-enabled) safe wrt SMP? Do we want a mutex instead?

(Note that this is used in many places)

Reply:
yeah, it is not used as a concurrency construct; it simply tells the compiler to not generate interrupt checks in this block for perf reasons.

Reply:
note that there is no point in putting a mutex here.
If you use the reader concurrently you are already shooting yourseld in the doot because it is non deterministic already.
So i dont see a reason to add a mutex as a lead boot around the user's foot!

Reply:
Added a small section in the docs documenting the thread-safety properties of Readers and Writers.
Basically common sense:
- don't read from two threads concurrently.
- don't write from two threads concurrently.
- sockets are full duplex, so you can read and write concurrently and also close from another thread.